server/src/main/java/com/cloud/resource/RollingMaintenanceManagerImpl.java

```diff
@@ -0,0 +1,488 @@
+                    s_logger.debug("Trying to set the host " + host.getId() + " into maintenance");
+                    resourceManager.maintain(host.getId());
+                    waitForHostInMaintenance(host.getId());
+
+                    result = performStageOnHost(host, Stage.Maintenance, timeout, payload, forced);
+                    if (!result.getLeft()) {
+                        if (forced) {
+                            String msg = "Maintenance script failed: " + result.getMiddle();
+                            hostsSkipped.add(new HostSkipped(host, msg));
+                            continue;
+                        }
+                        return new ImmutableTriple<>(result.getLeft(), result.getMiddle(), new Pair<>(hostsUpdated, hostsSkipped));
+                    }
+
+                    if (!resourceManager.cancelMaintenance(host.getId())) {
+                        String message = "Could not cancel maintenance on host " + host.getUuid();
+                        s_logger.error(message);
+                        throw new CloudRuntimeException(message);
+                    }
+                    Date endTime = new Date();
+
+                    HostUpdated hostUpdated = new HostUpdated(host, startTime, endTime);
+                    hostsUpdated.add(hostUpdated);
+
+                    result = performStageOnHost(host, Stage.PostMaintenance, timeout, payload, forced);
+                    if (!result.getLeft()) {
+                        if (forced) {
+                            String msg = "Post-maintenance script failed: " + result.getMiddle();
+                            hostsSkipped.add(new HostSkipped(host, msg));
+                            continue;
+                        }
+                        return new ImmutableTriple<>(result.getLeft(), result.getMiddle(), new Pair<>(hostsUpdated, hostsSkipped));
+                    }
+                }
+
+                // Enable back disabled cluster
+                if (cluster.getAllocationState() == Grouping.AllocationState.Disabled && disabledClusters.contains(clusterId)) {
+                    updateCluster(clusterId, "Enabled");
+                }
+            }
+        } catch (AgentUnavailableException | InterruptedException | CloudRuntimeException e) {
+            String err = "Error starting rolling maintenance: " + e.getMessage();
+            s_logger.error(err, e);
+            return new ImmutableTriple<>(false, err, new Pair<>(hostsUpdated, hostsSkipped));
+        } finally {
+            // Enable back disabled clusters
+            for (Long clusterId : disabledClusters) {
+                Cluster cluster = resourceManager.getCluster(clusterId);
+                if (cluster.getAllocationState() == Grouping.AllocationState.Disabled) {
+                    updateCluster(clusterId, "Enabled");
+                }
+            }
+        }
+        return new ImmutableTriple<>(true, "OK", new Pair<>(hostsUpdated, hostsSkipped));
+    }
```

Comment:
123 lines is way to long to get it. please refactor into screen size chucks (i'm on an old terminal, or at least my head is!)